README.MD
### 联合消息队列 UniJobsQueue：
#### 1、将支持多种驱动模式：redis（已支持）、rabbitMq、mysql
#### 2、延时队列支持自定义延时周期
#### 3、已拓展链式操作，快速引用
#### 4、将支持协程模式


CHANGELOG.md
CHANGELOG
---------

### v1.0.1, 2022.02.27

* README、CHANGELOG补充

### v1.0.0, 2022.02.27

* 第一个版本首次发布

* Started changelog


test/helpers.php
<?php

/*
 * demo helper UniJobsQueue.
 *
 * (c) Mansonwong <260945307@qq.com>
 *
 * This source file is subject to the Apache License that is bundled.
 */

require __DIR__.'/../vendor/autoload.php';

use Mansonwong\UniJobsQueue\UniJobsQueue;

if (! function_exists('jobs_queue')) {

    /**
     * Notes:获取实例类（可在任意框架自行实现读取初始化config）
     * @return UniJobsQueue
     */
    function jobs_queue()
    {
        $config = [];//your config
        return UniJobsQueue::getInstance($config);
    }
}


test/ConsumerTest.php
<?php

/*
 * This file is part of the UniJobsQueue.
 *
 * (c) Mansonwong <260945307@qq.com>
 *
 * This source file is subject to the Apache License that is bundled.
 */

require __DIR__.'/../vendor/autoload.php';

use Mansonwong\UniJobsQueue\UniJobsQueue;

/**
 * $config 配置参数，请参考 config/uniqueue.php，
 * 你可以简单做一个封装用在您的框架里面去，
 * 比如在laravel框架 可以使用服务提供者，助手函数、门面模式等方式做一个简单初始化配置封装
 * 或者不需要封装直接在你的业务类中初始化
 */
$config = [];


/****************** 非延时模式demo ******************/
$res1 = UniJobsQueue::getInstance($config)->model(UniJobsQueue::MODEL_LIST)->subscribe('demo',function ($rawData){
    // TODO: 执行成功请返回return true，执行失败返回return false,否则会认为job任务执行失败，队列回滚，用try catch
    $res = true;//假设任务执行成功true，失败为false
    if ($rawData){
        try {
            print_r($rawData);
            if (!$res){
                throw new \Exception('模拟执行失败！');
            }
            return true;
        } catch (\Exception $e) {
            print $e->getMessage();
            return false;
        }
    }
});

//判断是否有错误信息
if ($res1->getError()){
    echo '队列事务执行有错误';
    var_dump($res1->getError());
}



/****************** 延时模式demo ******************/
$res2 = UniJobsQueue::getInstance($config)->model(UniJobsQueue::MODEL_DELAY)->subscribe('demo',function ($rawData){

    // TODO:  执行成功请返回return true，执行失败返回return false,否则会认为job任务执行失败，队列回滚，用try catch
    $res = true;//假设任务执行成功true，失败为false
    if ($rawData){
        try {
            print_r($rawData);
            if (!$res){
                throw new \Exception('模拟执行失败！');
            }
            return true;
        } catch (\Exception $e) {
            print $e->getMessage();
            return false;
        }
    }
});

//判断是否有错误信息
if ($res2->getError()){
    echo '队列事务执行有错误';
    var_dump($res1->getError());
}

/**
 * 温馨提示：关于队列名，普通队列和延时队列模式的队列名同样是 demo ，其实不会冲突，按模式区分
 */

test/ProducerTest.php
<?php

/*
 * This file is part of the UniJobsQueue.
 *
 * (c) Mansonwong <260945307@qq.com>
 *
 * This source file is subject to the Apache License that is bundled.
 */

require __DIR__.'/../vendor/autoload.php';

use Mansonwong\UniJobsQueue\UniJobsQueue;


/**
 * $config 配置参数，请参考 config/uniqueue.php，
 * 你可以简单做一个封装用在您的框架里面去，
 * 比如在laravel框架 可以使用服务提供者，助手函数、门面模式等方式做一个简单初始化配置封装
 * 或者不需要封装直接在你的业务类中初始化
 */
$config = [];

/****************** 普通队列模式demo ******************/
UniJobsQueue::getInstance($config)
//    ->model(UniJobsQueue::MODEL_LIST)//不提供 则根据默认config/uniqueue.php 配置文件的 model 参数决定
    ->dispatch('demo',['hello'=>'我是普通队列'])//队列名 和 装载的数据
    ->attempts(10)//最大重试次数
    ->push();//执行推送消息(链式操作的最后)

/****************** 延时模式demo ******************/
UniJobsQueue::getInstance()
    ->model(UniJobsQueue::MODEL_DELAY)//设置为延时模式，不提供该参数将默认非延时模式
    ->dispatch('demo',['hello'=>'我是延时队列'])//队列名 和 数据
    ->delay(10)//首次延时
    ->attempts(10)//最大重试次数
    ->periods([10,30,60,180,1800,1800,3600,7200])//第二次之后的延时周期，如果不满足最大重试次数以数组最后一个元素为准
    ->push();//执行推送消息(链式操作的最后)

/**
 * 温馨提示：关于队列名，普通队列和延时队列模式的队列名同样是 demo ，其实不会冲突，按模式区分
 */

config/config.php
<?php

return [
    'model' => 'delay',//模式
    'delay' => 0,
    'attempts' => 0,
    'max_attempts' => 100,
    'driver' => 'redis',
    'delay_periods' => [30, 60, 180, 1800, 1800, 3600, 7200],
    /*
    |--------------------------------------------------------------------------
    | 队列驱动方式
    |--------------------------------------------------------------------------
    |
    | Drivers:  "database", "redis", "rabbitmq"
    |
    */
    'connections' => [
        'database' => [
            'driver' => 'database',
            'table' => 'jobs',//表
            'queue' => 'default',
            'retry_after' => 90,
        ],
        'redis' => [
            'driver' => 'redis',
            'host' => '127.0.0.1',
            'port' => 6379,
            'password' => '',
            'select' => 1,
            'timeout' => 0,
            'expire' => 0,
            'persistent' => true,
            'prefix' => 'jobs_queue',
        ],
        'rabbitmq' => [
            'host' => '127.0.0.1',
            'vhost' => 'my_vhost',
            'port' => 5672,
            'login' => 'admin',
            'password' => 'admin'
        ],

    ],

    /*
    | 失败的队列
    */
    'failed' => [
        'database' => 'redis',
        'table' => 'failed_jobs',
    ],

];


src/UniJobsQueue.php
<?php
/**
 *
 * 联合消息队列 UniJobsQueue 主逻辑
 * Created On 2022/2
 * Created By WangWenJun <260945307@qq.com>
 */

namespace Mansonwong\UniJobsQueue;

use Mansonwong\UniJobsQueue\driver\MysqlQueueDriver;
use Mansonwong\UniJobsQueue\driver\RabbitMqQueueDriver;
use  Mansonwong\UniJobsQueue\driver\RedisQueueDriver;

class UniJobsQueue implements UniJobsQueueInterface
{
    /**
     * 驱动实例
     * @var
     */
    protected static $driver;
    /**
     * 模式
     * @var
     */
    protected $model;
    /**
     * 延时模式
     */
    const MODEL_DELAY = 'delay';
    /**
     * 队列模式/可用于抢购
     */
    const MODEL_LIST = 'list';

    /**
     * 初始默认模式
     * @var
     */
    protected static $default_model = 'list';

    /**
     * 配置
     * @var null
     */
    protected static $config = null;
    /**
     * 驱动实体
     * @var
     */
    protected static $driverInstance;
    /**
     * 延时周期
     * @var array
     */
    protected $delay_periods = [];
    /**
     * 首次延时时间
     * @var int
     */
    protected $delay = 0;
    /**
     * 最大重试次数
     * @var int
     */
    protected $max_attempts = 10;
    /**
     * 队列名
     * @var null
     */
    protected $queue = null;
    /**
     * 组装的队列数据
     * @var array
     */
    protected $item = [];
    /**
     *  方法
     * @var null
     */
    protected $method = null;
    /**
     * 结果集
     * @var null
     */
    protected $result = null;
    /**
     * 当前类实例化
     * @var
     */
    protected static $instance;
    /**
     * 用于检测是否 设置模式
     * @var bool
     */
    protected $is_set_model = false;
    /**
     * @var mixed
     */
    public $error = false;


    public function __construct($config = [])
    {
        $this->initConfig($config);
    }


    /**
     * @return mixed
     */
    private function getDriver()
    {
        return $this->driver;
    }

    /**
     * @param mixed $driver
     */
    private function setDriver($driver)
    {
        $this->driver = $driver;
    }

    /**
     * @return mixed
     */
    private function getModel()
    {
        return $this->model;
    }

    /**
     * @param mixed $model
     */
    private function setModel($model)
    {
        $this->model = $model;
    }

    /**
     * @return array
     */
    private function getDelayPeriods()
    {
        return $this->delay_periods;
    }

    /**
     * @param array $delay_periods
     */
    private function setDelayPeriods($delay_periods)
    {
        $this->delay_periods = $delay_periods;
    }

    /**
     * @return int
     */
    private function getDelay()
    {
        return $this->delay;
    }

    /**
     * @param int $delay
     */
    private function setDelay($delay)
    {
        $this->delay = $delay;
    }

    /**
     * @return int
     */
    private function getMaxAttempts()
    {
        return $this->max_attempts;
    }

    /**
     * @param int $max_attempts
     */
    private function setMaxAttempts($max_attempts)
    {
        $this->max_attempts = $max_attempts;
    }

    /**
     * @return null
     */
    private function getQueue()
    {
        return $this->queue;
    }

    /**
     * @param null $queue
     */
    private function setQueue($queue)
    {
        $this->queue = $queue;
    }

    /**
     * @return array
     */
    private function getItem()
    {
        return $this->item;
    }

    /**
     * @param array $item
     */
    private function setItem($item)
    {
        $this->item = $item;
    }


    public static function getInstance($config=[])
    {
        if (self::$instance) {
            return self::$instance;
        }
        return new self($config);
    }

    /**
     * @return null
     */
    private function getMethod()
    {
        return $this->method;
    }

    /**
     * @param null $method
     */
    public function setMethod($method)
    {
        $this->method = $method;
    }


    /**
     * 设置队列参数
     * @param $queue
     * @param $payload
     * @return $this|mixed
     */
    public function dispatch($queue, $payload)
    {
        $this->setMethod('doDispatch');
        $this->setQueue($queue);
        $this->setItem(array(
            'queue' => $queue,
            'model' => self::$default_model,//默认为rpush rpop
            'attempts' => 0,
            'max_attempts' => $this->max_attempts,
            'payload' => $payload,
        ));
        return $this;
    }

    public function getCurrentDelayPeriod($delay_periods,$num)
    {
        if (isset($delay_periods[$num])){
            return $delay_periods[$num];
        }
        return end($delay_periods);
    }

    public function doDispatch()
    {
        if ($this->model == self::MODEL_DELAY) {
            $result = self::driverInstance()->pushDelay(self::getQueue(), self::getItem(), self::getDelay());
        } else {
            $result = self::driverInstance()->push(self::getQueue(), self::getItem());
        }
        $this->setResult($result);
        return $result;
    }

    public function doSubscribe()
    {
        if ($this->model == self::MODEL_DELAY) {
            //延迟模式
            $result = self::driverInstance()->popDelay($this->getQueue());
        } else {
            $result = self::driverInstance()->pop($this->getQueue());
        }
        return $result;
    }

    public function model($model)
    {
        $this->is_set_model = true;
        //延时模式
        $this->model = $model;
        return $this;
    }

    public function delay($delay = 0)
    {
        $this->setDelay($delay);
        $this->item['model'] = 'delay';
        $this->item['delay_periods'] = self::$config['delay_periods'];
        //延时模式
        $this->model = self::MODEL_DELAY;
        return $this;
    }

    public function periods($delay_periods)
    {
        $this->setDelayPeriods($delay_periods);
        return $this;
    }

    public function attempts(int $max_attempts)
    {
        $this->setMaxAttempts($max_attempts);
        $this->item['max_attempts'] = $max_attempts;
        return $this;
    }

    /**
     * @return mixed
     */
    public function push()
    {
        $fun = $this->getMethod();
        return $this->$fun();
    }


    /**
     * @param null $result
     */
    protected function setResult($result)
    {
        $this->result = $result;
    }


    private function callFunc()
    {
        $fun = $this->getMethod();
        return $this->$fun();
    }


    /**
     * 订阅/接收队列消息
     * @param $queue
     * @return $this|mixed
     */
    public function subscribe($queue, $callfunc)
    {
        if ($this->is_set_model == false) {
            echo '未设置model';
            die;
        }
        $this->setQueue($queue);
        $this->setMethod('doSubscribe');
        if ($this->model == self::MODEL_DELAY) {
            //延迟模式
            $rawData = self::driverInstance()->popDelay($queue);
        } else {

            $rawData = self::driverInstance()->pop($queue);
        }
        $jobres = call_user_func($callfunc, $rawData);
        if (!$jobres) {
            //如果回调函数返回false,表示job执行失败或者没有队列
            if ($rawData) {
                //如果失败回滚
                $roll = $this->reDispatch($rawData);
                if ($roll){
                    $this->setErrorInfo('回滚失败',['roll_status'=>$roll,'raw_data'=>$rawData]);
                }
            }
        } else {
//            echo '成功了！！！';
        }

        return $this;

    }


    /**
     * 重新进入对列
     * @param $item
     * @return mixed
     */
    public function reDispatch($item)
    {

        $this->model = $item['model'];
//        echo '模式为：' . $this->model;
        $queue = $item['queue'];
        $item['attempts']++;
        if ($item['attempts'] <= $item['max_attempts']) {
        }else{
            //推入失败队列 redis
            $queue = self::$config['failed']['table'] . ':' . $queue;
        }
        if ($this->model == self::MODEL_DELAY) {
            //延迟模式
            $reItem = array(
                'id' => $item['id'],
                'model' => $item['model'],
                'queue' => $item['queue'],
                'attempts' => $item['attempts'],
                'max_attempts' => $item['max_attempts'],
                'delay_periods' => $item['delay_periods'],
                'payload' => $item['payload'],
            );
            $delay = $this->getCurrentDelayPeriod($item['delay_periods'],$item['attempts']);
            return self::driverInstance()->pushDelay($queue, $reItem, $delay);
        } else {
            $reItem = array(
                'id' => $item['id'],
                'model' => $item['model'],
                'queue' => $item['queue'],
                'attempts' => $item['attempts'],
                'max_attempts' => $item['max_attempts'],
                'payload' => $item['payload'],
            );
            return self::driverInstance()->push($queue, $reItem);
        }
    }

    /**
     * 获取消息长度
     * @param $queue
     * @return mixed
     */
    public function size($queue)
    {
        if ($this->model == self::MODEL_DELAY) {
            //延迟模式
            return self::driverInstance()->sizeDelay($queue);
        } else {
            return self::driverInstance()->size($queue);
        }
    }

    /**
     * Notes:初始化配置
     * @param $userConfig
     * @return array|mixed|null
     * @throws \Exception
     * Author: Wangwenjun <260945307@qq.com>
     */
    private function initConfig($userConfig)
    {
        if (!self::$config) {
            $defaultConfig = $this->getConfigByFile();
            if ($userConfig) {
                self::$config = array_merge($defaultConfig, $userConfig);
            } else {
                self::$config = $defaultConfig;
            }
        }
        if (isset(self::$config['model'])){
            $this->model(self::$default_model);
            self::$default_model = self::$config['model'];//初始默认模式
        }
        return self::$config;
    }

    public static function driverInstance()
    {
        if (self::$driverInstance !== null) {
            return self::$driverInstance;
        }
        self::$driver = self::$config['driver'];
        switch (self::$driver) {
            case 'redis':
                self::$driverInstance = new RedisQueueDriver(self::$config['connections']['redis']);
                break;
            case 'mysql':
                self::$driverInstance = new MysqlQueueDriver(self::$config['connections']['mysql']);
                break;
            case 'rabbitmq':
                self::$driverInstance = new RabbitMqQueueDriver(self::$config['connections']['rabbitmq']);
                break;
            default:
                exit('driver error');
                break;
        }
        return self::$driverInstance;
    }

    private function getConfigByFile()
    {
        $config = include __DIR__ . '/../config/uniqueue.php';
        if (!$config) {
            throw new \Exception('config error');
        }
        return $config;
    }

    /**
     * Notes:设置错误信息
     * @param string $msg
     * @param array $data
     */
    private function setErrorInfo($msg='',$data=[])
    {
        $this->error = array(
            'msg'=>$msg,
            'data'=>$data,
        );
    }

    /**
     * Notes:获取错误信息
     * @return array|mixed
     */
    public function getError()
    {
        if (!$this->error){
            return [];
        }
        return $this->error;
    }


}


src/driver/RedisQueueDriver.php
<?php
namespace Mansonwong\UniJobsQueue\driver;

class RedisQueueDriver implements QueueDriverFactory
{
    const QUEUE_PREFIX = 'jobs_queue:';
    const DELAY_QUEUE_PREFIX = 'jobs_delay_queue:';
    const DELAY_QUEUE_TEMP_PREFIX = 'jobs_delay_queue_temp:';

    protected static $options = [
        'host' => '127.0.0.1',
        'port' => 6379,
        'password' => '',
        'select' => 0,
        'timeout' => 0,
        'expire' => 0,
        'persistent' => false,
        'prefix' => '',
    ];

    public function __construct($options = [])
    {
        if (!empty($options)) {
            self::$options = array_merge(self::$options, $options);
        }
    }

    public static $redis = null;
    protected static $redisServer = null;

    /**
     * Notes: 引用外部redis
     * @param $server
     * Author: Wangwenjun <260945307@qq.com>
     * Time: 2022/2/26
     */
    public static function setBackend($server)
    {
        self::$redisServer   = $server;
        self::$redis         = null;
    }

    public static function redis()
    {
        if (self::$redis !== null) {
            return self::$redis;
        }
        if (self::$redisServer){
            //如果有指定redis
            self::$redis = self::$redisServer;
            return self::$redis;
        }
        self::$redis = new \Redis();
        if (self::$options['persistent']) {
            self::$redis->pconnect(self::$options['host'], self::$options['port'], self::$options['timeout'], 'persistent_id_' . self::$options['select']);
        } else {
            self::$redis->connect(self::$options['host'], self::$options['port'], self::$options['timeout']);
        }
        if ('' != self::$options['password']) {
            self::$redis->auth(self::$options['password']);
        }
        if (0 != self::$options['select']) {
            self::$redis->select(self::$options['select']);
        }
        return self::$redis;
    }

    public function popDelay($queue)
    {
        $key = self::DELAY_QUEUE_PREFIX.$queue;
        $data = self::zRangeByScore($key);

        if (empty($data)) {
            return false;
        }
        //每次只取一条任务
        $item = $data[0];
        //立刻移除有序队列集中的一个成员，防止并发被多人抢到
        if (self::zRem($key,$item)) {
            $item = json_decode($item, true);
            //返回数据
            return $item;
        }
        return false;
    }

    public function push($queue, $item)
    {
        if (!isset($item['id'])){
            $item['id'] = self::generateJobId(); //这里跟那里一样
        }
        $now = time();
        $item['reserved_at'] = $now; //预约时间
        $encodedItem = json_encode($item);
        if ($encodedItem === false) {
            return false;
        }

        $length = self::redis()->rpush(self::QUEUE_PREFIX. $queue, $encodedItem);
        if ($length < 1) {
            return false;
        }
        return true;

    }

    public function pushDelay($queue, $item,$delay)
    {
        if (!isset($item['id'])){
            $item['id'] = self::generateJobId(); //这里跟那里一样
        }
        $now = time();
        $available_time = $now+$delay;
        $item['reserved_at'] = $now; //预约时间
        $item['available_at'] = $available_time; //计划运行时间
        $encodedItem = json_encode($item,JSON_UNESCAPED_UNICODE);
        if ($encodedItem === false) {
            return false;
        }
        $res =  self::redis()->zAdd(
            self::DELAY_QUEUE_PREFIX.$queue,
            $available_time,//以时间作为score，队列按时间从小到大排序
            $encodedItem
        );
        if ($res){
            return $item;
        }
        return false;
    }
    public static function zRangeByScore($key)
    {
        //以0和当前时间为区间，返回一条记录
        return self::redis()->zRangeByScore($key, 0, time(), array('limit' => array(0, 1)));
    }
    public static function zRem($key,$value)
    {
        return self::redis()->zRem($key, $value);
    }

    public function pop($queue)
    {
        $item = self::redis()->lpop(self::QUEUE_PREFIX.$queue);
        if(!$item) {
            return false;
        }
        return json_decode($item, true);
    }

    public function size($queue)
    {
        return self::redis()->llen(self::QUEUE_PREFIX.$queue);
    }

    public function sizeDelay($queue)
    {
        return self::zRangeByScore(self::DELAY_QUEUE_PREFIX.$queue);
    }

    public static function generateJobId()
    {
        return md5(uniqid('', true));
    }
}


config/uniqueue.php
<?php

return [
    'model' => 'list',//默认模式，非延时模式：list、 延时模式：delay
    'delay' => 0,//初始默认首次延时
    'max_attempts' => 100,//最大失败重试次数
    'driver' => 'redis',//驱动模式
    'delay_periods' => [30, 60, 180, 1800, 1800, 3600, 7200],
    /*
    |--------------------------------------------------------------------------
    | 队列驱动方式
    |--------------------------------------------------------------------------
    |
    | Drivers:  "mysql", "redis", "rabbitmq"
    |
    */
    'connections' => [
        'database' => [
            'driver' => 'mysql',
            'table' => 'jobs',//表
            'queue' => 'default',
            'retry_after' => 90,
        ],
        'redis' => [
            'driver' => 'redis',
            'host' => '127.0.0.1',
            'port' => 6379,
            'password' => '',
            'select' => 1,
            'timeout' => 0,
            'expire' => 0,
            'persistent' => true,
            'prefix' => 'jobs_queue',
        ],
        'rabbitmq' => [
            'host' => '127.0.0.1',
            'vhost' => 'my_vhost',
            'port' => 5672,
            'login' => 'admin',
            'password' => 'admin'
        ],

    ],

    /*
    | 失败的队列
    */
    'failed' => [
        'database' => 'redis',
        'table' => 'failed_jobs',
    ],

];
